File: #0200.number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {

        int size = grid.size();
        if (size == 0) {
            return 0;
        }

        int rowSize = grid[0].size();
        if (rowSize == 0) {
            return 0;
        }

        int result = 0;
        vector<vector<bool>> visited(size, vector<bool>(rowSize));
        vector<int> X{-1, 0, 1, 0};
        vector<int> Y{0, 1, 0, -1};
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < rowSize; ++j) {
                if (grid[i][j] == '0' || visited[i][j]) {
                    continue;
                }
                ++result;
                queue<int> q{{i * rowSize + j}};
                while (!q.empty()) {
                    int t = q.front();
                    q.pop();
                    for (int k = 0; k < 4; ++k) {
                        int x = t / rowSize + X[k];
                        int y = t % rowSize + Y[k];
                        if (x < 0 || x >= size || y < 0
                            || y >= rowSize || grid[x][y] == '0'
                            || visited[x][y]) {
                            continue;
                        }
                        visited[x][y] = true;
                        q.push(x * rowSize + y);
                    }
                }
            }
        }
        return result;
    }
};

File: #0200.number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    const x = [-1, 0, 0, 1];
    const y = [0, -1, 1, 0];
    const isCheck = (x, y) => {
        return x >= 0 && y >= 0 && x < grid.length
            && y < grid[0].length 
            && grid[x][y] === '1'
    };

    const isColor = (i, j) => {
        if (!grid[i][j]) {
            return;
        }

        grid[i][j] = '0';
        for (let k = 0 ; k < 4 ; k++ ) {
            const dx = x[k] + i;
            const dy = y[k] + j;
            if (isCheck(dx, dy) && grid[dx][dy] === '1') {  
                isColor(dx, dy);
            }
        }
    }

    let result = 0;
    for (let i = 0 ; i < grid.length ; i++) {
        for (let j = 0 ; j < grid[0].length ; j++) {
            if (grid[i][j] === '1') {  
                isColor(i, j);
                result++;
            }
        }
    }
    return result; 
};
